pkg/util/util.go

```diff
@@ -0,0 +1,283 @@
+func IsNilOrInvalidValue(v reflect.Value) bool {
+	return !v.IsValid() || (v.Kind() == reflect.Ptr && v.IsNil()) || IsValueNil(v.Interface())
+}
+
+// AppendToSlicePtr inserts value into parent which must be a slice ptr.
+func AppendToSlicePtr(parentSlice interface{}, value interface{}) error {
+	dbgPrint("AppendToSlicePtr slice=\n%s\nvalue=\n%v", pretty.Sprint(parentSlice), value)
+	pv := reflect.ValueOf(parentSlice)
+	v := reflect.ValueOf(value)
+
+	if !IsSliceInterfacePtr(parentSlice) {
+		return fmt.Errorf("AppendToSlicePtr parent type is %T, must be *[]interface{}", parentSlice)
+	}
+
+	pv.Elem().Set(reflect.Append(pv.Elem(), v))
+
+	return nil
+}
+
+func DeleteFromSlicePtr(parentSlice interface{}, index int) error {
+	dbgPrint("DeleteFromSlicePtr index=%d, slice=\n%s", index, pretty.Sprint(parentSlice))
+	pv := reflect.ValueOf(parentSlice)
+
+	if !IsSliceInterfacePtr(parentSlice) {
+		return fmt.Errorf("DeleteFromSlicePtr parent type is %T, must be *[]interface{}", parentSlice)
+	}
+
+	pvv := pv.Elem()
+	if pvv.Kind() == reflect.Interface {
+		pvv = pvv.Elem()
+	}
+
+	ns := reflect.AppendSlice(pvv.Slice(0, index), pvv.Slice(index+1, pvv.Len()))
+	pv.Elem().Set(ns)
+
+	return nil
+}
+
+func UpdateSlicePtr(parentSlice interface{}, index int, value interface{}) error {
+	dbgPrint("UpdateSlicePtr parent=\n%s\n, index=%d, value=\n%v", pretty.Sprint(parentSlice), index, value)
+	pv := reflect.ValueOf(parentSlice)
+	v := reflect.ValueOf(value)
+
+	if !IsSliceInterfacePtr(parentSlice) {
+		return fmt.Errorf("UpdateSlicePtr parent type is %T, must be *[]interface{}", parentSlice)
+	}
+
+	pvv := pv.Elem()
+	if pvv.Kind() == reflect.Interface {
+		pv.Elem().Elem().Index(index).Set(v)
+		return nil
+	}
+	pv.Elem().Index(index).Set(v)
+
+	return nil
+}
+
+// InsertIntoMap inserts value with key into parent which must be a map, map ptr, or interface to map.
+func InsertIntoMap(parentMap interface{}, key interface{}, value interface{}) error {
+	dbgPrint("InsertIntoMap key=%v, value=%s, map=\n%s", key, pretty.Sprint(value), pretty.Sprint(parentMap))
+	v := reflect.ValueOf(parentMap)
+	kv := reflect.ValueOf(key)
+	vv := reflect.ValueOf(value)
+
+	if v.Type().Kind() == reflect.Ptr {
+		v = v.Elem()
+	}
+	if v.Type().Kind() == reflect.Interface {
+		v = v.Elem()
+	}
+
+	if v.Type().Kind() != reflect.Map {
+		dbgPrint("error %v", v.Type().Kind())
+		return fmt.Errorf("InsertIntoMap parent type is %T, must be map", parentMap)
+	}
+
+	v.SetMapIndex(kv, vv)
+
+	return nil
+}
+
+func IsYAMLEqual(a, b string) bool {
+	if strings.TrimSpace(a) == "" && strings.TrimSpace(b) == "" {
+		return true
+	}
+	ajb, err := yaml.YAMLToJSON([]byte(a))
+	if err != nil {
+		dbgPrint("bad YAML in isYAMLEqual:\n%s", a)
+		return false
+	}
+	bjb, err := yaml.YAMLToJSON([]byte(b))
+	if err != nil {
+		dbgPrint("bad YAML in isYAMLEqual:\n%s", b)
+		return false
+	}
+
+	return string(ajb) == string(bjb)
+}
+
+func YAMLDiff(a, b string) string {
```

Comment:
In IsYAMLEqual function it do comparison by converting them to JSON first but in YAMLDiff it unmarshall to map[string]interface{} first and marshal then compare. What is the reason for difference behind them?

Reply:
The reason that YAMLDiff unmarshals and marshals is to make the field ordering the same in the yaml. If ordering is the same we can use text diff.Diff to get a nice diff output for the yaml comparison.